### ISSUE TITLE
BUG: `np.lib.index_tricks.as_strided` no longer a public function in numpy 2.0

### ISSUE THREAD

**Issue by david-zwicker** [numpy/numpy#26725]
### Describe the issue:

I used to use `np.lib.index_tricks.as_strided` to set the strides of an array, but this function is now private in numpy 2.0. However, the [documentation of the `.strides` attribute](https://numpy.org/devdocs/reference/generated/numpy.ndarray.strides.html) still suggest to use that function to set strides.

### Reproduce the code example:

```python
import numpy as np
np.lib.index_tricks.as_strided(np.arange(4))
```

### Error message:

```shell
raise AttributeError(
AttributeError: numpy.lib.index_tricks is now private. If you are using a public function, it should be available in the main numpy namespace, otherwise check the NumPy 2.0 migration guide.
```

### Python and NumPy Versions:

Python 3.12.3 and numpy 2.0.0

### Context for the issue:

I am passing memory between a numbs-compiled function and a numpy function using pointers (the context is indeed a bit complicated). Long story short, the function turning the data back into a numpy array receives the data_pointer from one side and knows the shape and strides from another side and tries to merge that information to re-create the numpy array.

**Comment by seberg** [comment]
Hmmm, the function always lived in `np.lib.stride_tricks.as_strided`.  The only reason it is available through that namespace is that other functions in that namespace were using it.

Unfortunately, I doubt there is anything to do.  One of the down-sides of having a non-clear API surface in `np.lib` historically.

**Comment by david-zwicker** [comment]
I understand the necessity of a clear API, and that this might break compatibility. However, it would then be great to adjust the documentation to tell users how to adjust the strides (or say that this is not possible with the public API).

**Comment by rkern** [comment]
> The only reason it is available through that namespace is that other functions in that namespace were using it.

Sort of. It was kind of intended as a kind of a gray area "it's here and you can use it (not just numpy internals), but its kind of buried down here because you can segfault if you're not careful". But over time, we started to try to have a more black-and-white view of being either completely public and completely private, and there's no room in that model for this kind of gray area.

**Comment by seberg** [comment]
@david-zwicker to be clear again, your code will work fine on both new and old NumPy versions if you just use:
```
from numpy.lib.stride_tricks import as_strided
# or
np.lib.stride_tricks.as_strided
```

The issue is a namespace confusion only, since the function was previously additionally in the `index_tricks` namespace (which was really just a private place to implement things that were public also in the main namespace, I believe).

@mtsokol maybe we should add it to the https://numpy.org/devdocs/numpy_2_0_migration_guide.html guide?  It actually would have been cute to include the URL in the `AttributeError`, but not sure it is worth doing that now.

**Comment by david-zwicker** [comment]
Ahh, now I understand. I guess was not reading the error message carefully enough and confused `stride_tricks` with `index_tricks`. I guess the issue can be closed then. Thank you very much!

**Comment by mtsokol** [comment]
`as_strided` always lived in `numpy.lib.stride_tricks` and was accidentally reexported in `numpy.lib.index_tricks`. During one of the cleanups we made sure each function is only present in one place. For `as_strided` it's `numpy.lib.stride_tricks`.

Across the NumPy public API (`np.core` and `np.lib` modules especially) there was quite a number of functions importable from multiple places. We removed duplicates and defined what is public and what is private, but we didn't provide a custom message for every removed duplicate across a codebase - therefore for `numpy.lib.index_tricks.as_strided` the error is just `AttributeError`.

**Comment by david-zwicker** [comment]
Yes, this makes perfect sense – I don't think anything needs to be changed (except my old code, which imported the function from the wrong module). Cleaning up the namespace will certainly help avoiding such issues in the future! Keep up the good work!

**Comment by mtsokol** [comment]
> @mtsokol maybe we should add it to the https://numpy.org/devdocs/numpy_2_0_migration_guide.html guide? It actually would have been cute to include the URL in the `AttributeError`, but not sure it is worth doing that now.

@seberg There was a lot of functions available in multiple modules, so tracking and documenting every one of them would clutter migration guide in my opinion.

**Comment by seberg** [comment]
> so tracking and documenting every one of them would 

Agreed, if it was just a few it might have made sense.  This one is one of the more common, "namespace only" functions, but I doubt it matters.  And it shouldn't be super common that it is imported from `index_tricks` rather than `stride_tricks` anyway.
